Bug: Missing peer dependencies

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Playground Link

_No response_

### Repro Code

```json
{
  "devDependencies": {
    "eslint": "9.4.0",
    "typescript": "5.4.5",
    "typescript-eslint": "8.0.0-alpha.25"
  }
}
```

```
Package @typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe] is requested to provide typescript by its descendants

@typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe]
└─ @typescript-eslint/typescript-estree@npm:8.0.0-alpha.25 [fd6b8] (via *)
   └─ ts-api-utils@npm:1.3.0 [e2004] (via >=4.2.0)

✘ Package @typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe] does not provide typescript.
```

```
Package @typescript-eslint/type-utils@npm:8.0.0-alpha.25 [47dfd] is requested to provide eslint by its descendants

@typescript-eslint/type-utils@npm:8.0.0-alpha.25 [47dfd]
└─ @typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe] (via ^8.57.0 || ^9.0.0)
   └─ @eslint-community/eslint-utils@npm:4.4.0 [fd6b8] (via ^6.0.0 || ^7.0.0 || >=8.0.0)

✘ Package @typescript-eslint/type-utils@npm:8.0.0-alpha.25 [47dfd] does not provide eslint.
```

```
Package typescript-eslint@npm:8.0.0-alpha.25 [14e3a] is requested to provide eslint by its descendants

typescript-eslint@npm:8.0.0-alpha.25 [14e3a]
├─ @typescript-eslint/eslint-plugin@npm:8.0.0-alpha.25 [a23fe] (via ^8.57.0 || ^9.0.0)
│  └─ @typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe] (via ^8.57.0 || ^9.0.0)
│     └─ @eslint-community/eslint-utils@npm:4.4.0 [fd6b8] (via ^6.0.0 || ^7.0.0 || >=8.0.0)
├─ @typescript-eslint/parser@npm:8.0.0-alpha.25 [a23fe] (via ^8.57.0 || ^9.0.0)
└─ @typescript-eslint/utils@npm:8.0.0-alpha.25 [a23fe] (via ^8.57.0 || ^9.0.0)
   └─ ...

✘ Package typescript-eslint@npm:8.0.0-alpha.25 [14e3a] does not provide eslint.
```

### ESLint Config

_No response_

### tsconfig

_No response_

### Expected Result

Installed without any warnings. 

### Actual Result

`yarn install` reports unmet peer dependencies. 

### Additional Info

_No response_

### Versions

<!-- Fill in the relevant versions and delete the others -->
| package | version |
| - | - |
| `typescript-eslint` | `8.0.0-alpha.25` |
| `TypeScript` | `5.4.5` |
| `ESLint` | `9.4.0` |

Hi,

Same problem here today but using NPM.

My dev dependencies versions:
```
"devDependencies": {
    "@typescript-eslint/eslint-plugin": "^7.12.0",
    "@typescript-eslint/parser": "^7.12.0",
    "eslint": "^9.4.0",
  }
```

When I try to run `npm Install`:
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: project-name@0.1.0
npm ERR! Found: eslint@9.4.0
npm ERR! node_modules/eslint
npm ERR!   dev eslint@"^9.4.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer eslint@"^8.56.0" from @typescript-eslint/parser@7.12.0
npm ERR! node_modules/@typescript-eslint/parser
npm ERR!   dev @typescript-eslint/parser@"^7.12.0" from the root project
npm ERR!   peer @typescript-eslint/parser@"^7.0.0" from @typescript-eslint/eslint-plugin@7.12.0
npm ERR!   node_modules/@typescript-eslint/eslint-plugin
npm ERR!     dev @typescript-eslint/eslint-plugin@"^7.12.0" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

I am getting exactly the same thing when I use yarn in pnpm mode
```
p25f29 → ✘ `@typescript-eslint/type-utils@npm:8.3.0` [bf79f] doesn't provide eslint to `@typescript-eslint/utils@npm:8.3.0` [bc0f5] and 1 other dependency
pdf5ff → ✘ `@typescript-eslint/utils@npm:8.3.0` [8e893] doesn't provide typescript to `@typescript-eslint/typescript-estree@npm:8.3.0` [1e2f5] and 1 other dependency
p25b1b → ✘ `@typescript-eslint/utils@npm:8.3.0` [bc0f5] doesn't provide typescript to `@typescript-eslint/typescript-estree@npm:8.3.0` [1e2f5] and 1 other dependency
```

I first started seeing this in `@typescript-eslint/eslint-plugin@npm:8.0.0`.

My solution to this was to add this to my .yarnrc.yml:
```yml
packageExtensions:
  "@typescript-eslint/type-utils@*":
    peerDependencies:
      eslint: "*"
```

<details>
<summary>
In my case I got the following error when trying to run eslint:
```
</summary>

```javascript
Error: Cannot find module 'eslint/use-at-your-own-risk'
Require stack:
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-utils-virtual-1639fa6ec9/package/dist/ts-eslint/eslint/LegacyESLint.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-utils-virtual-1639fa6ec9/package/dist/ts-eslint/ESLint.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-utils-virtual-1639fa6ec9/package/dist/ts-eslint/index.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-utils-virtual-1639fa6ec9/package/dist/index.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-type-utils-virtual-b8b27d4b16/package/dist/isTypeReadonly.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-type-utils-virtual-b8b27d4b16/package/dist/index.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-eslint-plugin-virtual-54e9de6f23/package/dist/util/misc.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-eslint-plugin-virtual-54e9de6f23/package/dist/util/index.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-eslint-plugin-virtual-54e9de6f23/package/dist/rules/adjacent-overload-signatures.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-eslint-plugin-virtual-54e9de6f23/package/dist/rules/index.js
- /home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-eslint-plugin-virtual-54e9de6f23/package/dist/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1145:15)
    at Module._load (node:internal/modules/cjs/loader:986:27)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/home/runner/work/node-stuffs/node-stuffs/node_modules/.store/@typescript-eslint-utils-virtual-1639fa6ec9/package/dist/ts-eslint/eslint/LegacyESLint.js:5:32)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:[10](https://github.com/scagood/node-stuffs/actions/runs/10188556353/job/28184924114?pr=283#step:6:11))
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:[12](https://github.com/scagood/node-stuffs/actions/runs/10188556353/job/28184924114?pr=283#step:6:13)33:19)
```

</details>

--- a/packages/type-utils/package.json
+++ b/packages/type-utils/package.json
@@ -51,6 +51,9 @@
     "debug": "^4.3.4",
     "ts-api-utils": "^1.3.0"
   },
+  "peerDependencies": {
+    "eslint": "^8.57.0 || ^9.0.0"
+  },
   "devDependencies": {
     "@jest/types": "29.6.3",
     "@typescript-eslint/parser": "8.0.1",

--- a/packages/typescript-eslint/package.json
+++ b/packages/typescript-eslint/package.json
@@ -56,6 +56,9 @@
     "@typescript-eslint/parser": "8.0.1",
     "@typescript-eslint/utils": "8.0.1"
   },
+  "peerDependencies": {
+    "eslint": "^8.57.0 || ^9.0.0"
+  },
   "devDependencies": {
     "@jest/types": "29.6.3",
     "downlevel-dts": "*",

--- a/packages/utils/package.json
+++ b/packages/utils/package.json
@@ -78,6 +78,11 @@
     "rimraf": "*",
     "typescript": "*"
   },
+  "peerDependenciesMeta": {
+    "typescript": {
+      "optional": true
+    }
+  },
   "funding": {
     "type": "opencollective",
     "url": "https://opencollective.com/typescript-eslint"

